### PR TITLE
Add more robust testing for BSON and bad auth sessions

### PR DESCRIPTION
**Pull request by jsflax**

### CODE REVIEW

File: Realm/ObjectServerTests/RLMBSONTests.mm [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+        @"int": @25,
+        @"int32": @5,
+        @"int64": @10,
+        @"double": @15.0,
+        @"decimal128": [[RLMDecimal128 alloc] initWithString:@"1.2E+10" error:nil],
+        @"minkey": [RLMMinKey new],
+        @"maxkey": [RLMMaxKey new],
+        @"date": [[NSDate alloc] initWithTimeIntervalSince1970: 500],
+        @"nestedarray": @[@[@1, @2], @[@3, @4]],
+        @"nesteddoc": @{@"a": @1, @"b": @2, @"c": @NO, @"d": @[@3, @4]},
```
Comment (leemaguire) [inline Realm/ObjectServerTests/RLMBSONTests.mm:115]: Should this not include `@"e" : @{@"foo": @"bar"}` as another nested element of the document?

File: Realm/RLMBSON.mm [hunk 2]
```diff
@@ -189,9 +189,9 @@ - (RLMBSONType)bsonType {
 }
 
 - (instancetype)initWithBsonBinary:(std::vector<char>)bsonBinary {
-    if ((self = [self initWithBase64EncodedData:[NSData dataWithBytesNoCopy:bsonBinary.data()
-                                                                     length:bsonBinary.size()
-                                                               freeWhenDone:NO] options:0])) {
+    if ((self = [NSData dataWithBytes:bsonBinary.data()
+                                     length:bsonBinary.size()
+                               ])) {
```
Comment (leemaguire) [inline Realm/RLMBSON.mm:194]: bad indent?

File: Realm/RLMBSON.mm [hunk 3]
```diff
@@ -316,11 +317,12 @@ Bson RLMConvertRLMBSONToBson(id<RLMBSON> b) {
...
             return util::none;
         case RLMBSONTypeBool:
-            return ((NSNumber *)b).boolValue;
+            return (bool)((NSNumber *)b).boolValue;
         case RLMBSONTypeDouble:
             return ((NSNumber *)b).doubleValue;
         case RLMBSONTypeBinary:
-            return (char**)((NSData *)b).bytes;
+            return std::vector<char>((char*)((NSData *)b).bytes,
+                                     ((char*)((NSData *)b).bytes) + (int)((NSData *)b).length);
```
Comment (leemaguire) [inline Realm/RLMBSON.mm:325]: bad indent

Comment (jsflax) [reply Realm/RLMBSON.mm:325]: this one is weirdly correct, but the above one is wrong

File: Realm/RLMBSON.mm (hunk 2 above)
Comment (jsflax) [reply Realm/RLMBSON.mm:194]: yep

File: Realm/ObjectServerTests/RLMBSONTests.mm (hunk 1 above)
Comment (jsflax) [reply Realm/ObjectServerTests/RLMBSONTests.mm:115]: It 's a nice to have.

**Pull request merged**